core/sphinx/packet_factory/packet_factory.go

```diff
@@ -0,0 +1,136 @@
+package packet_factory
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/katzenpost/katzenpost/core/crypto/rand"
+	"github.com/katzenpost/katzenpost/core/epochtime"
+	"github.com/katzenpost/katzenpost/core/pki"
+	"github.com/katzenpost/katzenpost/core/sphinx"
+
+	"github.com/katzenpost/katzenpost/core/sphinx/constants"
+	"github.com/katzenpost/katzenpost/core/sphinx/geo"
+	"github.com/katzenpost/katzenpost/core/sphinx/path"
+)
+
+// PacketFactory is used to compose Sphinx packets.
+type PacketFactory struct {
+	pathFactory *path.PathFactory
+	geo         *geo.Geometry
+	sphinx      *sphinx.Sphinx
+}
+
+// PacketFactoryOption is an option to the Sphinx packet factory.
+type PacketFactoryOption func(*PacketFactory)
+
+// WithPathFactory is used to set the path factory.
+func WithPathFactory(pathFactory *path.PathFactory) PacketFactoryOption {
+	return func(PacketFactory *PacketFactory) {
+		PacketFactory.pathFactory = pathFactory
+	}
+}
+
+func NewPacketFactory(geo *geo.Geometry, opts ...PacketFactoryOption) *PacketFactory {
+	factory := &PacketFactory{
+		pathFactory: new(path.PathFactory),
+		geo:         geo,
+	}
+	for _, opt := range opts {
+		opt(factory)
+	}
+	return factory
+}
+
+func (p *PacketFactory) ComposePacket(
+	doc *pki.Document,
+	srcId []byte,
+	srcMix *[32]byte,
+	dstId []byte,
+	dstMix *[32]byte,
+	surbID *[constants.SURBIDLength]byte,
+	message []byte) (packet []byte, surbKey []byte, rtt time.Duration, err error) {
```

Comment:
There seems to be some code cloning between this and ComposeSphinxPacket in `minclient/send.go` ?

Reply:
minclient is going away forever.